storage_service/locations/models/duracloud.py

```diff
@@ -129,16 +145,95 @@ def delete_path(self, delete_path):
+                root = etree.fromstring(response.content)
+                to_delete = [e.attrib['chunkId'] for e in root.findall('chunks/chunk')]
+                to_delete.append(delete_path + self.MANIFEST_SUFFIX)
+                LOGGER.debug('Chunks to delete: %s', to_delete)
+            else:
+                # File cannot be found - this may be a folder
+                to_delete = self._get_files_list(delete_path, show_split_files=True)
             # Do not support globbing for delete - do not want to accidentally
             # delete something
             for d in to_delete:
                 url = self.duraspace_url + urllib.quote(d)
                 response = self.session.delete(url)
 
+    def _download_file(self, url, download_path, expected_size=0, checksum=None):
+        """
+        Helper to download files from DuraCloud.
+
+        :param url: URL to fetch the file from.
+        :param download_path: Absolute path to store the downloaded file at.
+        :return: True on success, False if file not found
+        :raises: StorageException if response code not 200 or 404
+        """
+        LOGGER.debug('URL: %s', url)
+        response = self.session.get(url)
+        LOGGER.debug('Response: %s', response)
+        if response.status_code == 404:
+            # Check if chunked by looking for a .dura-manifest
+            manifest_url = url + self.MANIFEST_SUFFIX
+            LOGGER.debug('Manifest URL: %s', manifest_url)
+            response = self.session.get(manifest_url)
+            LOGGER.debug('Response: %s', response)
+            # No manifest - this file does not exist
+            if not response.ok:
+                return False
+
+            # Get chunks, expected size, checksum
+            root = etree.fromstring(response.content)
+            expected_size = int(root.findtext('header/sourceContent/byteSize'))
+            checksum = root.findtext('header/sourceContent/md5')
+            chunk_elements = [e for e in root.findall('chunks/chunk')]
+            # Download each chunk and append to original file
+            self.space._create_local_directory(download_path)
+            LOGGER.debug('Writing to %s', download_path)
+            with open(download_path, 'wb') as output_f:
+                for e in chunk_elements:
+                    # Parse chunk element
+                    chunk = e.attrib['chunkId']
+                    size = int(e.findtext('byteSize'))
+                    md5 = e.findtext('md5')
+                    # Download
+                    chunk_url = self.duraspace_url + urllib.quote(chunk)
+                    LOGGER.debug('Chunk URL: %s', chunk_url)
+                    chunk_path = chunk_url.replace(url, download_path)
+                    LOGGER.debug('Chunk path: %s', chunk_path)
+                    self._download_file(chunk_url, chunk_path, size, md5)
+                    # Append to output
+                    shutil.copyfileobj(open(chunk_path, 'rb'), output_f)
```

Comment:
Good catch. I wonder if that's why downloading chunked files takes up so much memory?